keps/sig-node/4639-oci-volume-source/README.md

```diff
@@ -0,0 +1,964 @@
+  guidance on best practices for users.
+
+## Design Details
+
+The new `VolumeSource` will be defined in the Kubernetes API, and the implementation will involve updating components (CRI, Kubelet)
+to support this source type. Key design aspects include:
+
+- API changes to introduce the new `VolumeSource` type.
+- Modifications to the Kubelet to handle mounting OCI images and artifacts.
+- Handling image pull secrets and registry authentication.
+- Reuse existing logic from ConfigMaps for:
+  - Determining the file location on the host filesystem.
+  - Managing garbage collection of unused images and artifacts.
+  - Handling updates to the OCI image or artifact, similar to how ConfigMaps update mounted files.
+
+The following code snippet illustrates the proposed API change:
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: example-pod
+spec:
+  volumes:
+  - name: oci-volume
+    oci:
+      image: "example.com/my-image:latest"
+  containers:
+  - name: my-container
+    image: busybox
+    volumeMounts:
+    - mountPath: /data
+      name: oci-volume
+```
+
+### Kubelet and Container Runtime Interface (CRI) support for OCI artifacts
+
+Kubelet and the Container Runtime Interface (CRI) currently handle OCI images. To support OCI artifacts,
+potential enhancements may be required:
+
+**Extended Media Type Handling:**
+   - Update Kubelet and CRI to recognize and handle new media types associated with OCI artifacts.
+   - Ensure that pulling and storing these artifacts is as efficient and secure as with OCI images.
+
+**Lifecycling and Garbage Collection:**
+   - Implement custom logic for managing the lifecycle of various OCI artifacts.
+
+**Artifact-Specific Configuration:**
+   - Introduce new configuration options or CRD to handle the unique requirements of different types of OCI artifacts.
+
+**Validation:**
+   - Extend validation and security checks to cover new artifact types.
+
+**Storage Optimization:**
+   - Develop optimized storage solutions tailored for different artifact types,
+     potentially integrating with existing storage solutions or introducing new mechanisms.
+
+#### kubelet
+
+Some parts of the existing kubelet code can be reused, for example:
+
+- The logic how to ensure that an image exists on the node:
+  https://github.com/kubernetes/kubernetes/blob/39c6bc3/pkg/kubelet/images/image_manager.go#L102
+- The retrieval of available secrets for a pod:
+  https://github.com/kubernetes/kubernetes/blob/39c6bc3/pkg/kubelet/kubelet_pods.go#L988
+
+We consider to refactor the logic into re-usable bits for both, the kubelet as
+well as the volume plugin, because volume plugins run in a dedicated go routine
+at the beginning of the kubelet initialization:
+
+https://github.com/kubernetes/kubernetes/blob/39c6bc3/pkg/kubelet/kubelet.go#L1633-L1634
+
+Specifying a pull policy will not be supported in the alpha implementation of the
```

Comment:
I think it should be discussed anyway, it seems an important detail, otherwise it will cause a performance impact, no?

Reply:
using Pull Always breaks Never pull (disconnected - ignore `latest` tag)

If Not Present is a policy that significantly reduces time to start a pod, some registry providers count checking manifests of an image as a full pull for the image. The time cost of making an authorized connection and checking the manifest tree for local existence of all sha's is, significant, esp. when the use count hits the user's limit at the registry.

Reply:
though: shouldn't the runtimes be optimizing and not repulling bits they already have? they have to do the auth check, and it does now rely on the registry always being up (which was something we steered away from in https://github.com/kubernetes/enhancements/issues/2535 ) but at least the runtimes should optimize so we're not actually pulling always if the bits are already on disk at the right location

Reply:
> shouldn't the runtimes be optimizing and not repulling bits they already have

Yes, that's typical.

Reply:
> though: shouldn't the runtimes be optimizing and not repulling bits they already have? they have to do the auth check, and it does now rely on the registry always being up (which was something we steered away from in #2535 ) but at least the runtimes should optimize so we're not actually pulling always if the bits are already on disk at the right location

Yes, of course. 

My point was about the time to auth connect and checking the manifests, by request, when the image is already present. There is usually no charge for "connect" .. the charge is for the manifest checking, even though the cost is expected to minimal with existing size limits on the number of annotations/size of config etc. Pulling the layer blobs is not needed when their manifest referenced SHAs are in cache.

Thus not having a policy at all for these new pulls, begs the question. Though with the appropriate container runtime config (a never pull disconnected state switch) and using the alg from the ensure secret KEP we can presume pull if not present and ensure at the container runtime.  This would be my suggestion.

Reply:
Fair and if we go back to the prior approach again.. note the nit to allow the CRI integrations to provide a suitable and safe alternative config.